src/main/scala/beam/agentsim/infrastructure/geozone/Hexagon.scala

```diff
@@ -0,0 +1,52 @@
+package beam.agentsim.infrastructure.geozone
+
+private[geozone] sealed trait Hexagon[+T] {
+  def index: GeoIndex
+  def totalNumberOfBuckets: Int
+  def totalNumberOfCoordinates: Int
+  def split(bucketsGoal: Int): Seq[HexagonBranch]
+}
+
+private[geozone] case class HexagonBranch(index: GeoIndex, children: IndexedSeq[Hexagon[_]])
+    extends Hexagon[HexagonBranch] {
+  override lazy val totalNumberOfCoordinates: Int = children.map(_.totalNumberOfCoordinates).sum
+
+  override lazy val totalNumberOfBuckets: Int = children.map(_.totalNumberOfBuckets).sum
+
+  def chooseOneToSplit(bucketsGoal: Int): Int = {
+    TopDownEqualDemandsSplitter.chooseOneToSplit(children, bucketsGoal, totalNumberOfBuckets)
+  }
+
+  override def split(bucketsGoal: Int): Seq[HexagonBranch] = {
+    val position = chooseOneToSplit(bucketsGoal)
+    val element = children(position)
+    element.split(bucketsGoal)
+  }
+
+}
+
+private[geozone] case class HexagonLeaf(
+  index: GeoIndex,
+  coordinates: Set[WgsCoordinate]
+) extends Hexagon[HexagonLeaf] {
+
+  override lazy val totalNumberOfCoordinates: Int = coordinates.size
+
+  override lazy val totalNumberOfBuckets: Int = 1
+
+  override def split(bucketsGoal: Int): Seq[HexagonBranch] = {
+    val resultIndex = H3Wrapper.getChildren(index)
+    val pointsAndNewIndexes: Map[GeoIndex, Set[WgsCoordinate]] = coordinates.toSeq
+      .map { point =>
+        val newIndex = H3Wrapper.getIndex(point, index.resolution + 1)
+        newIndex -> point
+      }
+      .groupBy(_._1)
+      .mapValues(_.map(_._2).toSet)
```

Comment:
nitpick: what is `_._1` or `_._2`. I'd suggest to use `groupBy` with unapply for tuple:
```scala
.groupBy { case (geoIndex, _) => geoIndex}
.mapValues(_.map { case (_, coord) => coord }.toSet)
```